text/0031-transactions-with-contingent-inputs.md

```diff
@@ -0,0 +1,394 @@
+    // Outputs that must appear in the transaction for a transaction using this input to be valid
+    repeated external.TxOut required_outputs = 1;
+    // A maximum tombstone block value for the transaction. If zero, it means no maximum is enforced.
+    fixed64 max_tombstone_block = 2;
+}
+```
+
+It may be hashed in a standard way using the `mc-crypto-digestible` hashing scheme (like everything else).
+
+This `required_outputs` are there to allow the main use-case of agreeing to a swap.
+The `max_tombstone_block` is there both, to allow that some of the required outputs can be fog outputs (since fog imposes tombstone block limits),
+and to allow that the party who builds and signs a `SignedContingentInput` can establish a "time to live" on the swap request.
+
+The `TxIn` object is modified, adding an Optional `ContingentInputRules` member.
+
+```
+message TxIn {
+    // "Ring" of inputs, one of which is actually being spent.
+    repeated TxOut ring = 1;
+
+    // Proof that each TxOut in `ring` is in the ledger.
+    repeated TxOutMembershipProof proofs = 2;
+    
+    // Any rules which must be satisfied by the Tx that contains this TxIn
+    ContingentInputRules contingent_input_rules = 3;
+}
+```
+
+Since the `TxIn` objects are under the `TxPrefix`, which gets hashed to form the `message` for the overall ringct
+signature, at this point the `ContingentInputRules` are "under the signature" for every MLSAG.
+
+https://github.com/mobilecoinfoundation/mobilecoin/blob/694fd2de00443338e6dfc63359c6178840cf7e52/transaction/core/src/validation/validate.rs#L310
+
+https://github.com/mobilecoinfoundation/mobilecoin/blob/694fd2de00443338e6dfc63359c6178840cf7e52/transaction/core/src/ring_signature/rct_bulletproofs.rs#L370
+
+However, this is not compatible with our desired use-case, where one user, who doesn't know the `TxPrefix` signs one of the MLSAGs,
+and another user fills in the rest of the transaction and signs the other MLSAGs.
+
+Therefore we propose that when signing `RingMLSAG`'s, there are two different cases, one for the situation where there are `ContingentInputRules`,
+and the other for the case where there are not `ContingentInputRules`.
+* When there are no `ContingentInputRules` on the input, the message to the MLSAG is the `extended_message_digest` the same as before: https://github.com/mobilecoinfoundation/mobilecoin/blob/694fd2de00443338e6dfc63359c6178840cf7e52/transaction/core/src/ring_signature/rct_bulletproofs.rs#L209
+* When there are contingent input rules present on the input, the message signed by the the MLSAG just consists of the digest of the `TxIn` but always omitting the `TxOutMembershipProof`, and including any `ContingentInputRules`.
+
+(The `TxOutMembershipProof` should not be signed because consensus may enforce that membership proofs have similar root membership elements. There is no problem to allow the final submitter of the
+transaction to update the root membership elements of these proofs.)
```

Comment:
There seems to be a lot of checking that we should maybe be doing here, is it possible to quickly sketch the math of this out in latex or in another form that could be checked?

Reply:
I am happy to try to add a section where we try to articulate the security property precisely. I would like to avoid adding a companion latex document if possible, if only because that's not currently part of the MCIP process. We could explore making tex files also part of the MCIP repo I guess.